gdrive/sheets_client.py

```diff
@@ -0,0 +1,657 @@
+        props = reply.get("addSheet").get("properties")
+        sheet_title_to_id[props.get("title")] = props.get("sheetId")
+
+    return sheet_title_to_id
+
+
+def export_df_to_gdrive_speadsheet(df: pd.DataFrame, sheets_id: str, title="Sheet1"):
+    """
+    Exports an entire pandas dataframe to a Google Sheets Object.
+
+    Args:
+        df (pandas.DataFrame): Tabular data to be exported to a spreadsheet
+        title (str): Title for the target spreadsheet to write the data to.
+            default: "Sheet1" default value for new Google Sheets sheets object
+
+    Returns:
+        Google Sheets API Response: RAW response to the write operation
+    """
+    body = {"values": df.values.tolist()}
+    result = (
+        sheets_service.spreadsheets()
+        .values()
+        .append(
+            spreadsheetId=sheets_id,
+            range="%s!A1" % (title),
+            valueInputOption="USER_ENTERED",
+            body=body,
+        )
+        .execute()
+    )
+    if "error" in result:
+        raise error.ExportError(result["error"]["message"])
+
+    return result
+
+
+# Project specific functions
+
+
+def do_create_pivot_tables(
+    df: pd.DataFrame, page_names: (str, str, str), names_to_id: dict, sheets_id: str
+):
+    # Make a dictionary mapping the name of the column to its index, useful for the pivot tables.
+    col_dict = {}
+    for idx, val in enumerate(df.iloc[0]):
+        col_dict[val] = idx
+
+    create_first_visit_pt(sheets_id, names_to_id[page_names[0]], col_dict)
+    log.info(
+        "Added 2 pivot tables to %s (%s)" % (page_names[0], names_to_id[page_names[0]])
+    )
+
+    create_session_start_pt(sheets_id, names_to_id[page_names[1]], col_dict)
+    log.info(
+        "Added 2 pivot tables to %s (%s)" % (page_names[1], names_to_id[page_names[1]])
+    )
+
+    create_clicks_pt(sheets_id, names_to_id[page_names[2]], col_dict)
+    log.info(
+        "Added pivot table to %s (%s)" % (page_names[2], names_to_id[page_names[2]])
+    )
+
+    create_feedback_pt(sheets_id, names_to_id[page_names[3]], col_dict)
+    log.info(
+        "Added pivot table to %s (%s)" % (page_names[3], names_to_id[page_names[3]])
+    )
+
+    update_cell_value(sheets_id, page_names[0], "A17", "Total First Visits")
+    update_cell_value(
+        sheets_id,
+        page_names[0],
+        "A18",
+        '=GETPIVOTDATA("SUM of eventCount",A1, "eventName", "first_visit") + GETPIVOTDATA("SUM of eventCount",F1, "eventName", "first_visit")',
+    )
+    log.info("Wrote totals to %s" % (page_names[0]))
+
+    update_cell_value(sheets_id, page_names[1], "A17", "Total Sessions")
+    update_cell_value(
+        sheets_id,
+        page_names[1],
+        "A18",
+        '=GETPIVOTDATA("SUM of eventCount",A1, "eventName", "session_start") + GETPIVOTDATA("SUM of eventCount",F1, "eventName", "session_start")',
+    )
+    log.info("Wrote totals to %s" % (page_names[1]))
+
+
+def create_first_visit_pt(sheets_id, page_id, col_dict):
```

Comment:
instead of having these functions with mostly static bodies perhapps is better to create constants and just refrence those constants in the calling function?

Reply:
Ill have to check, i think some of them might need one or two parameters. If so, id prefer to stay consistent with the placement of the JSON bodies. If possible, it does make a lot of sense to move them out of the function and into a constant.